### PR TITLE
Prevent contracts from owning Ether, Tokens and other contracts.

### PR DESCRIPTION
**Pull request by Recmo**
As shown in the [ERC23][1] token specification, accidents happen, and contracts end up owning things that they should not own.

In this PR a propose a set of inheritable contracts that prevent its descendants from owning Ether, Tokens and/or other Contracts—to the extend this is possible. When this is not possible, functions are provided to allow the owner to transfer ownership to itself.

[1]: https://github.com/Dexaran/ERC23-tokens#lost-tokens-held-by-contracts

### CODE REVIEW

**Comment by Recmo** [comment]
For the tests I wrote two helper functions:

* `toPromise`: Turns any `web3.*` function into a function returning a promise.
* `expectThrow`: Expects a given promise to throw an error indicating a throw statement in EVM. Unfortunately, it has false-positives for 'invalid jump' and 'out of gas' errors.

File: contracts/ownership/HasNoEther.sol [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+/// allow the owner to reclaim this ether.
+///
+/// @notice Ether can still be send to this contract by:
+///  * calling functions labeled `payable`
+///  * `selfdestruct(contract_address)`
+///  * mining directly to the contract address
+contract HasNoEther is Ownable {
+
+  /// Constructor that rejects incoming Ether
+  /// @dev The flag `payabe` is added so we can access `msg.value`
```
Comment (maraoz) [inline contracts/ownership/HasNoEther.sol]: typo here

Comment (Recmo) [inline contracts/ownership/HasNoEther.sol]: :+1:

**Comment by Recmo** [comment]
I'm a little bit worried about the `HasNoTokens` contract. The ERC23 `tokenFallback` function signature is by no means final. If it changes, we should revisit the contract and update it. Similarly, there is proposal being made for an `approveAndCall` function that is roughly similar. This one could also be implemented.

**Comment by maraoz** [comment]
@Recmo Let's merge this and we can update it once ERC23 and [HumanStandardToken](https://github.com/ConsenSys/Tokens/blob/4045d1de0f10b0ee2441220154fc8b524772e92f/Token_Contracts/contracts/HumanStandardToken.sol#L52) are more mature.

**Pull request merged**